Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1338 @@
+LL here means **L**eft-to-right parsing, **L**eftmost derivation, as they scan strings from left to right and recovers a derivation them from the left (see below).
+
+The number between the parentheses indicates how many look-ahead tokens the parser can see.
+A parser with a look-ahead of `k` can make decisions by looking `k` tokens (or characters) ahead.
+A look-ahead of `k=1` is often chosen for better performance, but popular parser combinatrs libraries offer an option to use an arbitrary `k` with the `try` function. A LL(k) parser is often enough to parse most grammars in programming languages.
+
+On the other hand, Happy is an LALR(1) parser generator, where the R means **R**ightmost derivation in reverse. An LALR(1) (Look-Ahead LR) parser generator is a more performant albeit less powerful variant of an LR(1) parser; see <a href=https://monlih.github.io/happy-docs/#_sec_conflict_tips>section 8.4.1</a> of the Happy User Guide.
+Happy also supports GLR (Generalized LR) parsing, which this tutorial won't cover, but you can find more about it in the <a href=https://monlih.github.io/happy-docs/#_sec_glr)>section 3</a> of the Happy User Guide.
+
+As an example, suppose we have the expression `1 + 2 + 3`, which is parsed as the following tree:
+
+```text
+            +
+          ↙   ↘
+        1       +
+              ↙   ↘
+            2       3
+```
+
+Which was generated from this grammar:
+
+```
+exp -> num
+exp -> num "+" exp
+num -> [0-9]+
+```
+
+It means that in an expression such as `1 + 2 + 3`, a LL parser would perform substitutions by visiting the leftmost nodes first, recovering the tokens in this order:
+
+<pre>
+<i>exp</i>
+<b><i>num</i> + exp</b>
+<b>1</b> + <i>exp</i>
+1 + <b><i>num</i> + exp</b>
+1 + <b>2</b> + <i>exp</i>
+1 + 2 + <b><i>num</i></b>
+1 + 2 + <b>3</b>
+</pre>
+
+n.b.: **Bold** means the result of the derivation, and <i>italic</i> means the derivation that will be expanded next.
+
+Whilst an LR parser would do the following:
+
+<pre>
+<i>exp</i>
+<b>num + <i>exp</i></b>
+num + <b>num + <i>exp</i></b>
+num + num + <b><i>num</i></b>
+num + <i>num</i> + <b>3</b>
+<i>num</i> + <b>2</b> + 3
+<b>1</b> + 2 + 3
+</pre>
+
+For more information on leftmost and rightmost derivations, see [LL and LR Parsing Demystified](https://blog.reverberate.org/2013/07/ll-and-lr-parsing-demystified.html).
+
+Another difference is that Alex is a lexer generator and Happy a parser generator.
+It means that instead of writing Haskell files (`.hs`) like you do with Megaparsec, you will write Alex (`.x`) and Happy (`.y`) files, which in turn will automatically generate Haskell files with implementations of the lexer and parser recognizing the described grammar.
+
+Below are some pros and cons of Happy in comparison to Megaparsec:
+
+Pros:
+* LR parsers are more powerful than LL parsers, allowing to parse more complex grammars with greater ease.
```

Comment:
I feel a bit like I've cornered myself since I now feel like I'm talking about something I don't completely understand (I am a victim of the Dunning-Kruger effect it seems).

I wanted to introduce some theory and terminology to make Happy easier to grasp, but also not go too much in-depth with it. I decided to just remove this point to avoid making it feel overwhelming.

To be honest, I feel like there are other parts that could be deleted as well, by this judgment, so this is something to think about.